File: web/eXoResources/src/main/webapp/javascript/portal/UIMaskWorkspace.js
/**
 * Copyright (C) 2009 eXo Platform SAS.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

/**
 * The mask layer, that appears when an ajax call waits for its result
 */

(function(base, uiMaskLayer, $) {

  eXo.portal.UIMaskWorkspace = {

    show : function(maskId, width, height) {
      this.maskWorkpace = document.getElementById(maskId);
      if (this.maskWorkpace) {
    	var content = $(this.maskWorkpace).find('.UIMaskWorkspace');
        if (width > -1) {
        	content.width(width);
        }

        document.body.style.overflow = 'hidden';
        if (eXo.portal.UIMaskWorkspace.maskLayer == null) {
          var maskLayer = uiMaskLayer.createMask("UIPortalApplication",
              this.maskWorkpace, 30);
          eXo.portal.UIMaskWorkspace.maskLayer = maskLayer;
        }
        this.maskWorkpace.style.margin = "auto";
        this.maskWorkpace.style.display = "block";
        $("#UIWorkingWorkspace").addClass("background");

        var browser = base.Browser;
        eXo.portal.UIMaskWorkspace.resetPositionEvt();
        browser.addOnResizeCallback('mid_maskWorkspace', eXo.portal.UIMaskWorkspace.resetPositionEvt);
      }
    },

    hide : function(maskId) {
      this.maskWorkpace = document.getElementById(maskId);
      if (eXo.portal.UIMaskWorkspace.maskLayer == undefined
          || !this.maskWorkpace) {
        return;
      }
      uiMaskLayer.removeMask(eXo.portal.UIMaskWorkspace.maskLayer);
      eXo.portal.UIMaskWorkspace.maskLayer = null;
      this.maskWorkpace.style.display = "none";
      document.body.style.overflow = 'auto';
      $("#UIWorkingWorkspace").removeClass("background");
    },

    /**
     * Resets the position of the mask calls eXo.core.uiMaskLayer.setPosition to
     * perform this operation
     */
    resetPositionEvt : function() {
      var maskWorkpace = eXo.portal.UIMaskWorkspace.maskWorkpace;
      if (maskWorkpace && (maskWorkpace.style.display == "block")) {
	    var maskHeight = $(maskWorkpace).height();
        var content = $(maskWorkpace).find('.UIMaskWorkspace');
        if (content.height() < maskHeight) {
          content.css('margin-top', (maskHeight - content.height()) / 2);
        } else {
          content.css('margin-top', 0);
        }
        try {
          uiMaskLayer.blockContainer = document
              .getElementById("UIPortalApplication");
          uiMaskLayer.object = maskWorkpace;
          uiMaskLayer.scrollCallback();
        } catch (e) {
        }
        maskWorkpace.style.position = 'fixed';
        maskWorkpace.style.top = 0;
      }
    }
  };

  return eXo.portal.UIMaskWorkspace;
})(base, uiMaskLayer, $);

File: webui/src/main/java/org/exoplatform/portal/webui/workspace/UIMaskWorkspace.java
/**
 * Copyright (C) 2009 eXo Platform SAS.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.exoplatform.portal.webui.workspace;

import org.exoplatform.webui.application.WebuiRequestContext;
import org.exoplatform.webui.config.annotation.ComponentConfig;
import org.exoplatform.webui.config.annotation.EventConfig;
import org.exoplatform.webui.core.UIComponent;
import org.exoplatform.webui.core.UIComponentDecorator;
import org.exoplatform.webui.event.Event;
import org.exoplatform.webui.event.Event.Phase;
import org.exoplatform.webui.event.EventListener;

/** Created by The eXo Platform SAS Mar 13, 2007 */
@ComponentConfig(id = "UIMaskWorkspace", template = "system:/groovy/portal/webui/workspace/UIMaskWorkspace.gtmpl", events = @EventConfig(phase = Phase.DECODE, listeners = UIMaskWorkspace.CloseActionListener.class, csrfCheck = false))
public class UIMaskWorkspace extends UIComponentDecorator {

    private int width_ = -1;

    private int height_ = -1;

    private boolean isShow = false;

    private String cssClasses = "";

    // TODO: Seems the isUpdated is never true
    private boolean isUpdated = false;

    public int getWindowWidth() {
        return width_;
    }

    public int getWindowHeight() {
        return height_;
    }

    public void setWindowSize(int w, int h) {
        width_ = w;
        height_ = h;
    }

    public boolean isShow() {
        return isShow;
    }

    public void setShow(boolean bln) {
        this.isShow = bln;
        if (bln == false) {
            isUpdated = false;
        }
    }

    public boolean isUpdated() {
        return isUpdated;
    }

    public void setUpdated(boolean bln) {
        // this.isUpdated = bln;
    }

    public String getCssClasses() {
        return cssClasses;
    }

    public void setCssClasses(String cssClasses) {
        this.cssClasses = cssClasses;
    }

    public void reset() {
        cssClasses = "";
        setWindowSize(-1, -1);
    }

    public <T extends UIComponent> T createUIComponent(Class<T> clazz, String configId, String id) throws Exception {
        T uicomponent = super.createUIComponent(clazz, configId, id);
        setUIComponent(uicomponent);
        return uicomponent;
    }

    public <T extends UIComponent> T createUIComponent(Class<T> clazz) throws Exception {
        return createUIComponent(clazz, null, null);
    }

    public UIComponent setUIComponent(UIComponent uicomponent) {
        UIComponent oldOne = super.setUIComponent(uicomponent);
        setShow(uicomponent != null);
        return oldOne;
    }

    public static class CloseActionListener extends EventListener<UIComponent> {
        public void execute(Event<UIComponent> event) throws Exception {
            UIMaskWorkspace uiMaskWorkspace = null;
            UIComponent uiSource = event.getSource();
            if (uiSource instanceof UIMaskWorkspace) {
                uiMaskWorkspace = (UIMaskWorkspace) uiSource;
            } else {
                uiMaskWorkspace = uiSource.getAncestorOfType(UIMaskWorkspace.class);
            }
            if (uiMaskWorkspace == null || !uiMaskWorkspace.isShow()) {
                return;
            }
            uiMaskWorkspace.setUIComponent(null);
            uiMaskWorkspace.reset();
            WebuiRequestContext rContext = event.getRequestContext();
            rContext.getJavascriptManager().require("SHARED/uiMaskWorkspace", "maskWS")
                    .addScripts("maskWS.hide('" + uiMaskWorkspace.getId() + "');");
            rContext.addUIComponentToUpdateByAjax(uiMaskWorkspace);
        }
    }
}
